src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs

```diff
@@ -201,7 +201,11 @@ public ICompilationTracker FreezePartialStateWithTree(SolutionState solution, Do
                         var oldTree = oldState.GetSyntaxTree(cancellationToken);
 
                         compilationPair = compilationPair.ReplaceSyntaxTree(oldTree, tree);
-                        inProgressProject = inProgressProject.UpdateDocument(docState, textChanged: false, recalculateDependentVersions: false);
+
+                        // We're changing the text of this document to something else entirely.  So we need to pass
+                        // `textChanged: true, recalculateDependentVersions: true` so that any interested parties that
+                        // compute anything that is dependent on changed text get the right values.
+                        inProgressProject = inProgressProject.UpdateDocument(docState, contentChanged: true);
```

Comment:
this is the crux of the fix.  it was possible to take a prior syntax tree for this doc and update ourselves with it, but then say there was no change, so top-level-versions might be incorrect.  We have a supposition this is what's leading to the nfw.